README.md
# delay-image

[![Build Status][ci-img]][ci]
[![BrowserStack Status][browserstack-img]][browserstack]

Delay image load until it’s in viewport.

## Install

```sh
npm install delay-image --save
```

## Usage

By default, it will look for `data-src` attribute on image element and add `src`
attribute when image has been successfully loaded.

```html
<img id="jackie" data-src="jackie.jpg" />
```

```js
import delayImage from 'delay-image';

delayImage(document.querySelector('#jackie'), {
	onEnter: (element) => {
		// Image in viewport!
	}
});
```

## API

### delayImage(element, options)

Delay image load until it’s in viewport.

#### element

Type: `Element`

Image element.

#### options

Type: `Object`

| Property              | Type       | Default value                                 | Description                                                                                                                                                                                                                                          |
| --------------------- | ---------- | --------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `threshold`           | `number`   | `0`                                           | Positive value in pixels which will signal plugin to check for image presence earlier in document.                                                                                                                                                   |
| `scrollResizeHandler` | `Function` | `300`                                         | Window scroll and resize event handler. Useful if you want to use [throttle or debounce methods](https://github.com/niksy/element-within-viewport#throttle-debounce-scroll-resize) on those events. Should return new handler (original or wrapped). |
| `imageSource`         | `string`   | `element.getAttribute('data-src')`            | Image URL to load.                                                                                                                                                                                                                                   |
| `onEnter`             | `Function` | `() => {}`                                    | Callback to execute if image is within viewport (useful for loader initialization).                                                                                                                                                                  |
| `onSuccess`           | `Function` | `(element) => { element.src = imageSource; }` | Callback to execute if image has been successfully loaded. **If you define this callback, you need to add `src` attribute yourself.**                                                                                                                |
| `onFail`              | `Function` | `() => { ... }`                               | Callback to execute if image has failed to load.                                                                                                                                                                                                     |

### instance.destroy()

Destroy instance.

## Browser support

Tested in Edge 15, Chrome 72 and Firefox 65, and should work in all modern
browsers
([support based on Browserslist configuration](https://browserslist.dev/?q=bGFzdCAzIG1ham9yIHZlcnNpb25zLCBzaW5jZSAyMDE5LCBlZGdlID49IDE1LCBub3QgaWUgPiAw)).

## Test

For automated tests, run `npm run test:automated` (append `:watch` for watcher
support).

## License

MIT © [Ivan Nikolić](http://ivannikolic.com)

<!-- prettier-ignore-start -->

[ci]: https://travis-ci.com/niksy/delay-image
[ci-img]: https://travis-ci.com/niksy/delay-image.svg?branch=master
[browserstack]: https://www.browserstack.com/
[browserstack-img]: https://www.browserstack.com/automate/badge.svg?badge_key=MGV1ZTVGN0tFSHU5alpRUW54SFZwcnh1aDZrUHVOc3AySDdUenYyQkdHYz0tLW9vUjVjeG4rbU14YjNpbXpYU2tlV0E9PQ==--2683e49bc19f3e9bf2dbd4c65fa7a3a8e6809318

<!-- prettier-ignore-end -->


CHANGELOG.md
# Changelog

## [Unreleased][]

## [3.0.0][] - 2021-05-05

### Changed

-   Update dependencies
-   Make module ESM only

### Removed

-   **Breaking**: Drop IE support, supported browser is Edge 15+

## [2.0.0][] - 2020-02-20

### Changed

-   Abstract scroll and resize event handling, so events could be wrapped with
    more functionality (e.g. throttle)
-   Lock `image-promise` to `6.0.2` since we need IE9 support

### Added

-   `scrollResizeHandler` property

### Removed

-   `debounce` property

## [1.0.0][] - 2019-01-18

### Added

-   Refactored implementation

<!-- prettier-ignore-start -->

[1.0.0]: https://github.com/niksy/delay-image/tree/v1.0.0
[2.0.0]: https://github.com/niksy/delay-image/tree/v2.0.0

<!-- prettier-ignore-end -->

[unreleased]: https://github.com/niksy/delay-image/compare/v3.0.0...HEAD
[3.0.0]: https://github.com/niksy/delay-image/tree/v3.0.0


karma.conf.js
'use strict';

const path = require('path');
const fs = require('fs');
const {
	assert: assertModulePath,
	url: urlModulePath
} = require('node-libs-browser');

let config;

const isCI =
	typeof process.env.CI !== 'undefined' && process.env.CI !== 'false';
const isPR =
	typeof process.env.TRAVIS_PULL_REQUEST !== 'undefined' &&
	process.env.TRAVIS_PULL_REQUEST !== 'false';
const local = !isCI || (isCI && isPR);

const port = 0;

if (local) {
	config = {
		browsers: ['Chrome']
	};
} else {
	config = {
		hostname: 'bs-local.com',
		browserStack: {
			username: process.env.BROWSER_STACK_USERNAME,
			accessKey: process.env.BROWSER_STACK_ACCESS_KEY,
			startTunnel: true,
			project: 'delay-image',
			name: 'Automated (Karma)',
			build: 'Automated (Karma)'
		},
		customLaunchers: {
			'BS-Chrome': {
				base: 'BrowserStack',
				browser: 'Chrome',
				'browser_version': '72',
				os: 'Windows',
				'os_version': '7',
				project: 'delay-image',
				build: 'Automated (Karma)',
				name: 'Chrome'
			},
			'BS-Firefox': {
				base: 'BrowserStack',
				browser: 'Firefox',
				'browser_version': '65',
				os: 'Windows',
				'os_version': '7',
				project: 'delay-image',
				build: 'Automated (Karma)',
				name: 'Firefox'
			},
			'BS-Edge15': {
				base: 'BrowserStack',
				browser: 'Edge',
				'browser_version': '15',
				os: 'Windows',
				'os_version': '10',
				project: 'delay-image',
				build: 'Automated (Karma)',
				name: 'Edge15'
			}
		},
		browsers: ['BS-Chrome', 'BS-Firefox', 'BS-Edge15']
	};
}

module.exports = function (baseConfig) {
	baseConfig.set({
		basePath: '',
		frameworks: ['mocha', 'fixture'],
		files: [
			'test/**/*.css',
			'test/**/*.html',
			{
				pattern: 'test/**/*.jpg',
				watched: false,
				included: false,
				served: true,
				nocache: false
			},
			'test/**/.webpack.js'
		],
		exclude: [],
		preprocessors: {
			'test/**/*.html': ['html2js'],
			'test/**/.webpack.js': ['webpack', 'sourcemap']
		},
		reporters: ['mocha', 'coverage-istanbul'],
		port: port,
		colors: true,
		logLevel: baseConfig.LOG_INFO,
		autoWatch: false,
		client: {
			captureConsole: true
		},
		browserConsoleLogOptions: {
			level: 'log',
			format: '%b %T: %m',
			terminal: true
		},
		browserNoActivityTimeout: 60000,
		webpack: {
			mode: 'none',
			devtool: 'inline-source-map',
			resolve: {
				fallback: {
					assert: assertModulePath,
					url: urlModulePath
				}
			},
			module: {
				rules: [
					{
						test: /\.js$/,
						exclude: /node_modules/,
						use: [
							{
								loader: 'babel-loader'
							}
						]
					},
					{
						test: /\.js$/,
						exclude: /(node_modules|test)/,
						enforce: 'post',
						use: [
							{
								loader: 'istanbul-instrumenter-loader',
								options: {
									esModules: true
								}
							}
						]
					}
				]
			}
		},
		coverageIstanbulReporter: {
			dir: path.join(__dirname, 'coverage/%browser%'),
			fixWebpackSourcePaths: true,
			reports: ['html', 'text'],
			thresholds: {
				global: JSON.parse(
					fs.readFileSync(path.join(__dirname, '.nycrc'), 'utf8')
				)
			}
		},
		singleRun: true,
		concurrency: Infinity,
		...config
	});
};
